app/components/machine/driver-hwclouds/component.js

```diff
@@ -0,0 +1,377 @@
+      // parser.pathname; // => "/pathname/"
+      // parser.search;   // => "?search=test"
+      // parser.hash;     // => "#hash"
+      // parser.host;     // => "example.com:3000"
+      return parser;
+
+    }
+    return {
+      endpoint: proxyUrl,
+      url: path,
+      host: domain,
+      qp: qp
+    };
+  },
+
+  apiRequest: function(url, data) {
+    this.set('itemsLoading', true);
+
+    return fetch(url, data).then((xhr) => {
+
+      if (xhr.status >= 200 && xhr.status < 300) {
+
+        this.set('itemsLoading', false);
+
+        return xhr.json();
+      } else {
+
+        this.set('itemsLoading', false);
+
+        return Ember.RSVP.reject();
+      }
+    });
+  },
+
+  describeRequest: function(url, method, t, body) {
+    return {
+      url: `${url.endpoint.substr(-1) === '/' ? url.endpoint.replace(/\/$/, '') : url.endpoint}?${url.qp}`,
+      urlToEncode: url.url.substr(-1) === '/' ? url.url : `${url.url}/`,
+      method: method,
+      headers: {
+        'x-sdk-date':    t, // long date
+        'host':          url.host,
+        'content-type':  '',
+      },
+      queryString: url.qp,
+      body: body,
+      params: {}
+    };
+  },
+
+  signAuth: function(credentials, req, time) {
+    var request = req;
+    var addAuthorization = (credentials) => {
+      let r =                     request;
+
+      r.params.Timestamp =              time.format('YYYYMMDDTHHmmss') + 'Z';
+      r.params.ShortTimestamp =         time.format('YYYYMMDD');
+      r.params.SignatureMethod =        'SDK-HMAC-SHA256';
+      r.params.AWSAccessKeyId =         credentials.accessKeyId;
+      r.params.Signature =              signature(credentials);
+      r['Content-Length'] =             r.body.length;
+      r.headers['content-type'] +=      ' rancher:';
+      r.headers['x-api-auth-header'] =  this.buildAuth(r.params);
+
+      return r;
+    };
+
+    function signature(credentials) {
+      var canonicalRequest =    buildCanonicalRequest(request);
+      var hashedSigingString =  AWS.util.crypto.sha256(canonicalRequest, 'hex');
+      var credScope =           credentialScope(request.params.ShortTimestamp, 'eu-de', 'ecs');
+      var sts =                 stringToSign(hashedSigingString, credScope, request.params.Timestamp);
+      var key =                 generateSigningKey(credentials.secretAccessKey, 'eu-de', 'ecs', request.params.ShortTimestamp);
+
+      return signStringToSign(sts, key);
+    }
+
+    function generateSigningKey(sk, rn, sn, t) {
+      var kDate =     AWS.util.crypto.hmac(`SDK${sk}`, t, 'buffer');
+      var kRegion =   AWS.util.crypto.hmac(kDate, rn, 'buffer');
+      var kService =  AWS.util.crypto.hmac(kRegion, sn, 'buffer');
+      var kSigning =  AWS.util.crypto.hmac(kService, 'sdk_request', 'buffer');
+      return kSigning;
+    }
+
+    function signStringToSign(stringToSign, signingKey) {
+      return AWS.util.crypto.hmac(signingKey, stringToSign, 'hex');
```

Comment:
Can this just be inlined into `signature()`, two things that sound like stringToSign kept confusing me.